OAI OpenAIRE: addressing issue 3097 with access rights

paulo-graca opened the pull request:
## References
* Fixes #3097

## Description
This pull request addresses issue 3097 by ensuring that only values ending with "access" can be considered for datacite:rights.

## Instructions for Reviewers
This is a simple change of openaire4.xls file that validates if the value ends with "access" keyword. It's not a perfect solution, but intends to address cases where dc:rights have multiple values and has at least one "open access" (as an example).

In order to test it you first need to ensure you have OAI enabled:
`oai.enabled = true`

also ensure you have content on your OAI interface, you can index it using:
`[/dspace/bin/]dspace oai import -c`

then access any item on your OAI interface within the openaire context:
https://yourserver.com/server/oai/openaire4?verb=GetRecord&metadataPrefix=oai_openaire&identifier=oai:dev5.rcaap.pt:123456789/1

Commit:
addressing issue 3097

Commit:
also add support for oaire:file @accessRightsURI

uofmsean commented on dspace/config/crosswalks/oai/metadataFormats/oai_openaire.xsl:
```diff
@@ -526,19 +526,30 @@
    <!-- datacite:rights -->
    <!-- https://openaire-guidelines-for-literature-repository-managers.readthedocs.io/en/v4.0.0/field_accessrights.html -->
     <xsl:template match="doc:element[@name='dc']/doc:element[@name='rights']/doc:element" mode="datacite">
+        <xsl:variable name="rightsValue" select="doc:field[@name='value']/text()"/>
```
There is still an issue here since the select will only result in the first dc.rights value for a particular language being tested later in resolveRightsURI. In testing I added dc.rights[en_US]="Some test value" followed by dc.rights[en_US]="open access" and the datacite:rights element in the output was missing. Reversing the value order produced the correct result.

tdonohue commented on dspace/config/crosswalks/oai/metadataFormats/oai_openaire.xsl:
```diff
@@ -526,19 +526,30 @@
+        <!-- this conditions ensures what is referred in issue: #3097 -->
+        <!-- it's a solution to ensure that only values ended with "access" -->
+        <!-- can be used as datacite:rights -->
+        <xsl:if test="ends-with($lc_rightsValue,'access')">
```
I'm a little worried about this as a solution.  First, does OpenAIREv4 require this field to be English language at all times?  If so, should we assume that any value ending in "access" is a valid OpenAIREv4 value, or should we specifically look for one of the four values listed at https://openaire-guidelines-for-literature-repository-managers.readthedocs.io/en/v4.0.0/field_accessrights.html

I'm not against the approach taken in this PR, but I just wonder if it solves the problem or not.  It looks, to me, like this value is *not* required to be English language, see: http://vocabularies.coar-repositories.org/documentation/access_rights/   It also looks like it's only allowed to be one of 4 different values.

uofmsean replied:
@tdonohue, I agree will what you are saying. Since [OpenAIRE 3 Access Level](https://guidelines.openaire.eu/en/latest/literature/field_accesslevel.html) is taken from dc:rights, that could be why [OpenAIRE 4 Access Rights](https://openaire-guidelines-for-literature-repository-managers.readthedocs.io/en/v4.0.0/field_accessrights.html) is taken from the same metadata field. As you mentioned, the better approach may be to use a different metadata field. @paulo-graca, should [openaire4-types.xml](https://github.com/DSpace/DSpace/blob/main/dspace/config/registries/openaire4-types.xml) be extended to include datacite:rights? If so, then that might be the preferred metadata field to use. The legacy use of dc:rights certainly does complicate matters given the large variety of content that can be in that metadata field.

In my initial [comment](https://github.com/DSpace/DSpace/issues/3097#issuecomment-747027882) I also suggested checking for all possible exact values since checking for ' access' at the end was an imperfect solution. It was a quick-win in our environment though (considering our existing metadata) and simplified the XSLT.

As for supporting multiple languages, what is the correct approach in DSpace in the context of OAI crosswalks?

paulo-graca replied:
As is, like with [OpenAIRE 3 support](https://github.com/DSpace/DSpace/blob/main/dspace/config/crosswalks/oai/transformers/openaire.xsl#L31-L53), there is a pre-transformation process to normalize/translate existing `dc:rights` values to OpenAIRE 4 [supported values](https://openaire-guidelines-for-literature-repository-managers.readthedocs.io/en/v4.0.0/field_accessrights.html#definition-and-usage-instruction) that will only be used in this OAI context, doesn't affect whatever the repository already has stored.

I understand the need to detached it from English language and the easier solution could be to create a specific metadata field (we proposed that initially with OpenAIRE4 schema but agreed to use as much existing fields as possible), or even store the PURL value in `dc:rights`. But that would also have implications at the front-end and the end-user expectations, like:

- the value wouldn't be displayed out-of-the-box if we use a different field. Changes to DSpace+Angular would be required to enabled it to be "OpenAIRE 4" aware. And the requirement is kind of the opposite, DSpace 7 should be ready, but OpenAIRE4 should be an opt in feature. The admin must [install it](https://wiki.lyrasis.org/display/DSDOC7x/OpenAIRE4+Guidelines+Compliancy).
- using PURL/URI in `dc:rights`, without no changes to Angular, wouldn't give much to users. We could use [pipes](https://angular.io/guide/pipes) for translation, but with that, some business logic would be brought to the front-end and a full repository `dc:rights` translation process would be required to normalize all pre-existing metadata values. And considering the multiplicity of possibilities for `dc:rights` usage within DSpace community, I don't think that it's feasible. For instance in Portuguese the translation for "open access" is "acesso aberto" and we use and validate "openAccess" or "info:eu-repo/semantics/openAccess" for a `dc:rights` value for our repositories. We also have a translation mechanism at the presentation layer to display it differently to the user (with open access's iconography). But this is a specific implementation for RCAAP (that represents only 50% of all repositories in Portugal).

That said, I think the solution in place is the best possible for what it intends to achieve (despite some required adjustments). It doesn't mess with what the repository already has; It's an opt in feature and flexible enough to be extended/changed by anyone who looks to enhance it.

paulo-graca replied:
@tdonohue please check this transformation: https://github.com/DSpace/DSpace/blob/main/dspace/config/crosswalks/oai/transformers/openaire4.xsl#L79-L114 it's a pre normalization for work. That's why you will have "access" English wording.

tdonohue replied:
@paulo-graca : Oh, I see now. I didn't realize that normalization is happening to use the English term "access"..  Could we reference that normalization from this section of the `oai_openaire.xsl` file then?  Since it's being pre-normalized in this way, I'm OK with this approach.  It just wasn't immediately evident to me...which is why a comment here could help!  The comment could just says something like:

`We are checking to ensure that only values ending in "access" can be used as datacite:rights.  This is a valid solution as we pre-normalize dc.rights values in openaire4.xsl to end in the term "access" according to COAR Controlled Vocabulary`

tdonohue reviewed (changes requested):
@paulo-graca and @uofmsean : To me, this approach doesn't seem to fully solve the problem.  It seems like we are assuming that this field will always have English language text (which is not a valid assumption).  While I understand the current mapping is not ideal either, it seems like this PR only makes minor improvements for English-language installations of DSpace.

I'm not sure what I'd recommend exactly here, but it seems like we either should check for specific controlled vocabulary options (in multiple languages), or perhaps use a different metadata field which is *specific* to this OpenAIREv4 set of values.

Commit:
comments improvement and fix dc.right processing

paulo-graca commented:
@tdonohue I've added your comment (it was better than the one I've first added). And @uofmsean, I've fixed the issue with multiple dc:rights. Thank you so much for testing it!

tdonohue reviewed (approved):
👍 Thanks @paulo-graca ! This looks correct to me now.
